contracts/PooledStaking.sol

```diff
@@ -0,0 +1,593 @@
+      rewards[rewardCount - 1].next = rewardCount;
+    } else {
+      // otherwise this is the only unprocessed reward and it should be the first one
+      firstReward = rewardCount;
+    }
+
+    // update counter
+    ++rewardCount;
+  }
+
+  function hasPendingActions() public view returns (bool) {
+    return hasPendingBurns() || hasPendingDeallocations() || hasPendingRewards();
+  }
+
+  function processPendingActions() public {
+
+    while (true) {
+
+      Burn storage burn = burns[firstBurn];
+      Deallocation storage deallocation = deallocations[firstDeallocation];
+      Reward storage reward = rewards[firstReward];
+
+      bool canBurn = burn.burnedAt != 0;
+      bool canDeallocate = deallocation.deallocateAt != 0 && deallocation.deallocateAt <= now;
+      bool canReward = reward.rewardedAt != 0;
+
+      if (!canBurn && !canDeallocate && !canReward) {
+        // everything is processed
+        break;
+      }
+
+      if (
+        canBurn &&
+        (!canDeallocate || burn.burnedAt < deallocation.deallocateAt) &&
+        (!canReward || burn.burnedAt < reward.rewardedAt)
+      ) {
+
+        // O(n*m)
+        if (!_processFirstBurn()) {
+          emit ActionStatus(false);
+          return;
+        }
+
+        continue;
+      }
+
+      if (
+        canDeallocate &&
+        (!canReward || deallocation.deallocateAt < reward.rewardedAt)
+      ) {
+
+        // deallocation gas limit check here
+//        uint gas = gasleft();
+//        gas = gas - gasleft();
+//        emit ActionStatus(true);
+
+        // O(1)
+        _processFirstDeallocation();
+        continue;
+      }
+
+      // O(n)
+      if (!_processFirstReward()) {
+        emit ActionStatus(false);
+        return;
+      }
+    }
+
+    // everything is processed!
+    emit ActionStatus(true);
+  }
+
+  function hasPendingBurns() public view returns (bool) {
+    return burns[firstBurn].burnedAt != 0;
+  }
+
+  function _processFirstBurn() internal returns (bool) {
+
+    Burn storage burn = burns[firstBurn];
+    Contract storage _contract = contracts[burn.contractAddress];
+    address contractAddress = burn.contractAddress;
+
+    uint burnAmount = burn.amount <= _contract.staked ? burn.amount : _contract.staked;
+    uint burned = 0;
+    uint newContractStake = 0;
+
+    for (uint i = 0; i < _contract.stakers.length; i++) {
+
+      Staker storage staker = stakers[_contract.stakers[i]];
+      uint oldAllocation = staker.allocations[contractAddress];
+
+      // formula: staker_burn = staker_allocation / total_contract_stake * contract_burn
+      // reordered for precision loss prevention
+      uint stakerBurn = oldAllocation.mul(burnAmount).div(_contract.staked);
+      uint newStake = staker.staked.sub(stakerBurn);
+      staker.allocations[contractAddress] = oldAllocation.sub(stakerBurn);
+
+      // reduce other contracts' stakes if needed
+      for (uint j = 0; j < staker.contracts.length; j++) {
```

Comment:
i think we should consider enforcing a low upper bound here from a feature perspective here similar to the one resulting from the percentage limitation.